api/app/Models/PoolCandidate.php

```diff
@@ -1066,62 +1070,79 @@ public function computeAssessmentStatus()
                 continue;
             }
 
-            if ($hasEssentialSkillsToAssess) {
+            foreach ($step->poolSkills as $poolSkill) {
+                $result = $stepResults->firstWhere('pool_skill_id', $poolSkill->id);
+                $decision = $result?->assessment_decision;
 
-                // Check assessed essential skills on this step
-                $essentialSkillAssessments = $stepResults
-                    ->filter(function ($result) {
-                        return $result->poolSkill?->type === PoolSkillType::ESSENTIAL->name &&
-                            $result->assessment_result_type !== AssessmentResultType::EDUCATION;
-                    });
+                if ($poolSkill->type === PoolSkillType::ESSENTIAL->name) {
+                    if (! $result || is_null($result->assessment_decision)) {
+                        $hasToAssess = true;
 
-                if ($essentialSkillAssessments->isEmpty()) {
-                    $hasToAssess = true;
-                }
+                        continue;
+                    }
 
-                foreach ($essentialSkillAssessments as $essentialSkillAssessment) {
-                    $decision = $essentialSkillAssessment->assessment_decision;
+                    if ($decision === AssessmentDecision::UNSUCCESSFUL->name) {
+                        $hasFailure = true;
+                        break;
+                    }
+
+                    if ($decision === AssessmentDecision::HOLD->name) {
+                        $hasOnHold = true;
 
-                    switch ($decision) {
-                        case null:
+                        continue;
+                    }
+                } else {
+                    if ($poolSkill->skill->category === SkillCategory::TECHNICAL->name) {
```

Comment:
Given how much trouble we had wrapping our heads around this, this block could use a comment explaining why its here, I think 😛

Reply:
And/or maybe just leave this whole block of pseudocode in the function's docblock to summarize the logic?
```
foreach step in pool->assessmentSteps
  foreach skill in assessmentStep->skills:
    result = find matching assessment result
    if skill is essential:
      if result is UNSUCCESSFUL, THEN mark UNSUCCESSFUL and exit loop
      if result is HOLD THEN mark HOLD and continue loop (to look for failures)
      if result is null or undecided, THEN mark TO ASSESS and continue loop (to look for failures)
    else if skill is asset:
      if skill is Technical AND user did not claim skill, THEN skip and continue loop
      ELSE if null or undecided THEN mark TO ASSESS and continue loop (to look for essential failures)
      ELSE mark nothing and continue, since the result doesn't actually matter
  AND if step is Application Assessment then repeat the Essential switch statement education assessment result  
  stepStatus is first of UNSUCCESSFUL, TO ASSESS, HOLD, and else QUALIFIED  
```

Reply:
Yeah, that makes sense. I hope I got it right in [document some trickier parts of logic](https://github.com/GCTC-NTGC/gc-digital-talent/pull/10925/commits/eee4591cb9a605be75160add3f222e69a6f92905)